cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java

```diff
@@ -0,0 +1,210 @@
+
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.datasetFramework = datasetFramework;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), new TransactionSystemClientAdapter(txClient),
+        NamespaceId.SYSTEM, ImmutableMap.of(), null, null, messagingContext)),
+      org.apache.tephra.RetryStrategies.retryOnConflict(20, 100)
+    );
+
+    // Adding all metadata related subscribers
+    this.subscriberServices = Collections.singletonList(new DataAccessLineageSubscriber());
+  }
+
+  /**
+   * Sets the {@link DatasetId} for the {@link LineageDataset}. This method is only for testing.
+   */
+  @VisibleForTesting
+  public void setLineageDatasetId(DatasetId lineageDatasetId) {
+    this.lineageDatasetId = lineageDatasetId;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    // Start all subscriber services. All of them has no-op in start, so they shouldn't fail.
+    Futures.successfulAsList(subscriberServices.stream().map(s -> {
+      LOG.debug("Starting metadata subscriber {}", s.getClass().getSimpleName());
+      return s.start();
+    }).collect(Collectors.toList())).get();
+
+    LOG.debug("All metadata subscribers started");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    // This never throw
+    Futures.successfulAsList(subscriberServices.stream().map(s -> {
+      LOG.debug("Stopping metadata subscriber {}", s.getClass().getSimpleName());
+      return s.stop();
+    }).collect(Collectors.toList())).get();
+
+    for (Service service : subscriberServices) {
+      // The service must have been stopped, and calling stop again will just return immediate with the
+      // future that carries the stop state.
+      try {
+        service.stop().get();
+      } catch (ExecutionException e) {
+        LOG.warn("Exception raised when stopping service {}", service.getClass().getSimpleName(), e.getCause());
+      }
+    }
+
+    LOG.debug("All metadata subscribers stopped");
+  }
+
+  /**
+   * A message subscriber for consuming lineage messages from the {@link Constants.Lineage#TOPIC} configuration
+   * and writes to {@link LineageDataset}.
+   */
+  private final class DataAccessLineageSubscriber extends AbstractMessagingSubscriberService<DataAccessLineage> {
+
+    DataAccessLineageSubscriber() {
+      super(NamespaceId.SYSTEM.topic(cConf.get(Constants.Lineage.TOPIC)),
+            false, cConf.getInt(Constants.Lineage.EVENT_FETCH_SIZE),
+            cConf.getLong(Constants.Lineage.EVENT_POLL_DELAY_MILLIS), retryStrategy,
+            baseMetricsContext.childContext(ImmutableMap.of(
+              Constants.Metrics.Tag.TOPIC, cConf.get(Constants.Lineage.TOPIC),
+              Constants.Metrics.Tag.CONSUMER, "lineage.writer"
+            )));
```

Comment:
nit: better to keep it in previous line for readability

Reply:
?? You mean the closing bracket? I align it with the `childContext` call as it starts from there.